src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_inline_self_ref.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+use Symfony\Component\DependencyInjection\Argument\RewindableGenerator;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\LogicException;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\ParameterBag\FrozenParameterBag;
+
+/**
+ * This class has been auto-generated
+ * by the Symfony Dependency Injection Component.
+ *
+ * @final since Symfony 3.3
+ */
+class Symfony_DI_PhpDumper_Test_Inline_Self_Ref extends Container
+{
+    private $parameters;
+    private $targetDirs = array();
+
+    public function __construct()
+    {
+        $this->services = array();
+        $this->normalizedIds = array(
+            'app\\foo' => 'App\\Foo',
+        );
+        $this->methodMap = array(
+            'App\\Foo' => 'getFooService',
+        );
+
+        $this->aliases = array();
+    }
+
+    public function getRemovedIds()
+    {
+        return array(
+            'Psr\\Container\\ContainerInterface' => true,
+            'Symfony\\Component\\DependencyInjection\\ContainerInterface' => true,
+        );
+    }
+
+    public function compile()
+    {
+        throw new LogicException('You cannot compile a dumped container that was already compiled.');
+    }
+
+    public function isCompiled()
+    {
+        return true;
+    }
+
+    public function isFrozen()
+    {
+        @trigger_error(sprintf('The %s() method is deprecated since Symfony 3.3 and will be removed in 4.0. Use the isCompiled() method instead.', __METHOD__), E_USER_DEPRECATED);
+
+        return true;
+    }
+
+    /**
+     * Gets the public 'App\Foo' shared service.
+     *
+     * @return \App\Foo
+     */
+    protected function getFooService()
+    {
+        $b = new \App\Bar();
+        $a = new \App\Baz($b);
+
+        $this->services['App\Foo'] = $instance = new \App\Foo($a);
+
+        $b->foo = $instance;
+
+        $a->bar = $b;
```

Comment:
shouldn't we set this property earlier, to have the property set at the time of `new \App\Foo($a)` ?

Reply:
It's important to see deps as late as possible, so that the consumer see them as populated as possible. The reason is that code usually expect it this way. The typical exemple is doctrine's configuration object, which must be fully defined before injection, so that consumers can use it immediately.